### ISSUE TITLE
Rewriting automata

### ISSUE THREAD

**Issue by Robby-the-paladin** [OnionGrief/Chipollino#318]
Написать модуль для загрузки автоматов в интерактивный конвертер:

- [ ] Реализовать функционал загрузки пользовательского ввода автоматов в объекты Чиполлино.
- [ ] Добавить поддержку данного функционала в интерпретатор.

**Comment by Robby-the-paladin** [comment]
Грамматика:
node_id -> identifier(unicode::alnum / lit_c<'_'> / lit_c<'-'>)
cell_id -> digits<>.no_leading_zero()
transition -> LEXY_LIT("eps") | unicode::alnum | (LEXY_LIT("&") >> p<cell_id>)
stmt -> p<node_id> + p<node_id> + p<transition>
memory_state -> LEXY_LIT("o") | LEXY_LIT("c")
memory_cell -> p<cell_id> >> p<memory_state>
memory_lists -> while_(p<memory_cell>)
MFA_edge -> p<stmt> + p<memory_lists>
FA_edge -> p<stmt>
state_label -> LEXY_LIT("label") >> (LEXY_LIT("=") + p<node_id>)
terminal_mark -> LEXY_LIT("terminal")
initial_mark -> LEXY_LIT("initial_state")
state_description -> p<node_id> + partial_combination(p<state_label>, p<terminal_mark>, p<initial_mark>)
states -> list(p<state_description>, sep) + LEXY_LIT("...")
MFA -> LEXY_LIT("{") + p<states> + list(p<MFA_edge>, sep) + LEXY_LIT("}")
FA -> LEXY_LIT("{") + p<states> + list(p<FA_edge>, sep) + LEXY_LIT("}")
production -> LEXY_LIT("MFA") >> p<MFA> | LEXY_LIT("FA") >> p<FA>

**Comment by Robby-the-paladin** [comment]
Пример автомата для FA:
```
FA {
    ab label = baobab terminal initial_state;
    aba terminal
    ...
    aba bab v;
    bab ab o;
    ab aba c
}
```
Пример автомата для MFA:
```
MFA {
    ab label = baobab terminal initial_state;
    aba terminal
    ...
    aba bab v 12 o;
    bab ab eps 12 c;
    ab aba &12
}
```

**Comment by Robby-the-paladin** [comment]
Грамматика:
1. production   -> "MFA" MFA | "FA" FA
2. MFA          -> "{" states MFA_edges "}"
3. FA           -> "{" states FA_edges "}"
4. states       -> state_description "..." | states state_description "..."
5. state_description -> node_id variants
6. variants     -> "label" "=" node_id | "terminal" | "initial_state"
7. MFA_edges     -> MFA_edge MFA_edges | ε
8. FA_edges      -> FA_edge FA_edges | ε
9. MFA_edge      -> stmt memory_lists
10. FA_edge      -> stmt
11. stmt         -> node_id node_id transition
12. memory_lists -> memory_cell memory_lists | ε
13. memory_cell  -> cell_id memory_state
14. memory_state -> "o" | "c"
15. node_id  -> (любая строка)
16. cell_id      -> (любое число)
17. transition   -> "eps" | (любая буква или цифра) | "&" cell_id

**Comment by TonitaN** [comment]
Так, а где интеграция в генератор случайных объектов? Причём на базе вашего же языка теперь уж надо.

**Comment by Robby-the-paladin** [comment]
https://github.com/OnionGrief/Chipollino/issues/318#issuecomment-1920621658 Отдельный от парсера генератор уже залит в main, в последнем коммите(https://github.com/OnionGrief/Chipollino/commit/4d699605ccde3c1c9a2d8dfb1a7cfa70f388319b) сделал встроенный в лексер, но я не уверен, что это архитектурно правильно. Генерировать напрямую из правил lexy не получится, так что в последнем коммите я просто добавил методы generate для структур грамматики.

**Comment by TonitaN** [comment]
Мне вот вообще не видно по коду, что генератор генерирует только корректные MFA (т.е. те, у которых нет записи внутри той же ячейки или чтения внутри записи в ячейку). И такие условия негоже хардкодить, а нужно привязывать к типу и вызывать по нему для каждого из классов автоматов как метод для графа (выбирающий, можно ли построить выбранный вами переход, или ограничивающий возможные метки на переходах некоторым собственным пулом значений для выбора). И так для DFA проверять детерминизм, а для MFA как минимум корректность обработки памяти.  

По-хорошему, этот генератор должен бы накладываться на абстрактные методы построения графов по заданным параметрам (ветвление, длина, достижимость), и именно это и было первоначальной нашей целью, но уж ладно, времени на это не осталось. Но ограничение значений генератора для построения только корректных автоматов сделать необходимо.
